cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogFileManager.java

```diff
@@ -0,0 +1,209 @@
+   * @param eventTimestamp timestamp from logging event
+   * @return LogFileOutputStream output stream to the log file
+   * @throws IOException if there is exception while getting location or while writing meta data
+   */
+  public LogFileOutputStream getLogFileOutputStream(LogPathIdentifier logPathIdentifier,
+                                                    long eventTimestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = outputStreamMap.get(logPathIdentifier);
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createOutputStream(logPathIdentifier, eventTimestamp);
+    } else {
+      // rotate the file if needed (time has passed)
+      logFileOutputStream = rotateOutputStream(logFileOutputStream, logPathIdentifier, eventTimestamp);
+    }
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createOutputStream(final LogPathIdentifier identifier,
+                                                 long timestamp) throws IOException {
+    TimeStampLocation location = createLocation(identifier);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location.getTimeStamp(), location.getLocation());
+    } catch (Throwable e) {
+      // delete created file as there was exception while writing meta data
+      Locations.deleteQuietly(location.getLocation());
+      throw new IOException(e);
+    }
+
+    LOG.info("Created Avro file at {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(
+      location.getLocation(), schema, syncIntervalBytes, new Closeable() {
+      @Override
+      public void close() throws IOException {
+        outputStreamMap.remove(identifier);
+      }
+    });
+
+    outputStreamMap.put(identifier, logFileOutputStream);
+    return logFileOutputStream;
+  }
+
+
+  private TimeStampLocation createLocation(LogPathIdentifier logPathIdentifier) throws IOException {
+    // we throw IOException only when getLocation fails, however if there is any error during createNew,
+    // we retry after sleeping for a milli second as we use current timestamp for fileName.
+    // this retry should succeed on any potential conflicts, though the likelihood of conflict is very small.
+    TimeStampLocation location = getLocation(logPathIdentifier);
+    LOG.trace("Trying to create new file at Location {}", location);
+    try {
+      // NOTE : we are creating a file before writing meta data, so if there are simultaneous write to same folder from
+      // multiple instance of log appender, they wouldn't overwrite the meta data and would fail early,
+      // so we can retry with different timestamp for file creation.
+      // however there is a possibility for the log.saver could crash after file was created
+      // but before meta data was written, log clean up should handle this scenario.
+      // log cleanup shouldn't rely on metadata table for cleaning up old files.
+      location.getLocation().createNew();
```

Comment:
`createNew` returns boolean to indicate success/failure, instead of throwing exception.